File: audit/backend/internal/audit/policy_engine.go
package audit

import (
	"fmt"
	"net/url"
	"sort"
	"strings"
)

// PolicyInput contains all the data needed for policy evaluation.
type PolicyInput struct {
	RiskScore        float64
	RiskSignals      []string
	QrUrls           []string
	Entities         Entities
	DeclaredLandingURL string
}

// Entities holds extracted entities from the ad creative.
type Entities struct {
	URLs       []string
	QrPayloads []string
	RiskTerms  []string
}

// PolicyResult is the outcome of running the policy engine.
type PolicyResult struct {
	Decision     string   // PASS, REJECT, or MANUAL_REVIEW
	MatchedRules []string
	Explanation  string
}

type rule struct {
	ID       string
	Name     string
	Check    func(input PolicyInput) bool
	Decision string
	Priority int // lower = higher priority
}

var rules = []rule{
	{
		ID:   "R001",
		Name: "QR URL mismatch with declared landing page",
		Check: func(input PolicyInput) bool {
			if len(input.QrUrls) == 0 {
				return false
			}
			declared, err := url.Parse(input.DeclaredLandingURL)
			if err != nil {
				return false
			}
			declaredDomain := declared.Hostname()
			for _, qrURL := range input.QrUrls {
				parsed, err := url.Parse(qrURL)
				if err != nil {
					// Unparseable QR URL is a mismatch
					return true
				}
				if parsed.Hostname() != declaredDomain {
					return true
				}
			}
			return false
		},
		Decision: "REJECT",
		Priority: 1,
	},
	{
		ID:   "R002",
		Name: "Contains false profit guarantees",
		Check: func(input PolicyInput) bool {
			if !containsSignal(input.RiskSignals, "risk_terms_detected") {
				return false
			}
			falseProfitTerms := []string{"guaranteed returns", "100% profit", "risk-free"}
			for _, term := range input.Entities.RiskTerms {
				for _, fp := range falseProfitTerms {
					if term == fp {
						return true
					}
				}
			}
			return false
		},
		Decision: "REJECT",
		Priority: 1,
	},
	{
		ID:   "R003",
		Name: "Suspicious redirect chain detected",
		Check: func(input PolicyInput) bool {
			return containsSignal(input.RiskSignals, "suspicious_redirect")
		},
		Decision: "REJECT",
		Priority: 2,
	},
	{
		ID:   "R004",
		Name: "High risk domain detected",
		Check: func(input PolicyInput) bool {
			return containsSignal(input.RiskSignals, "high_risk_domain")
		},
		Decision: "REJECT",
		Priority: 2,
	},
	{
		ID:   "R005",
		Name: "QR code present - needs review",
		Check: func(input PolicyInput) bool {
			return containsSignal(input.RiskSignals, "qr_code_found")
		},
		Decision: "MANUAL_REVIEW",
		Priority: 5,
	},
	{
		ID:   "R006",
		Name: "Short link usage detected",
		Check: func(input PolicyInput) bool {
			return containsSignal(input.RiskSignals, "short_link_detected")
		},
		Decision: "MANUAL_REVIEW",
		Priority: 5,
	},
	{
		ID:   "R007",
		Name: "Telegram handle does not match project",
		Check: func(input PolicyInput) bool {
			return containsSignal(input.RiskSignals, "telegram_mismatch")
		},
		Decision: "MANUAL_REVIEW",
		Priority: 6,
	},
	{
		ID:   "R008",
		Name: "Contains wallet connect / claim language",
		Check: func(input PolicyInput) bool {
			walletTerms := []string{"connect wallet", "wallet connect", "claim", "claim now"}
			for _, term := range input.Entities.RiskTerms {
				lower := strings.ToLower(term)
				for _, wt := range walletTerms {
					if lower == wt {
						return true
					}
				}
			}
			return false
		},
		Decision: "MANUAL_REVIEW",
		Priority: 7,
	},
}

// EvaluatePolicy runs all policy rules against the input and returns the result.
func EvaluatePolicy(input PolicyInput) PolicyResult {
	var matched []rule

	for _, r := range rules {
		if r.Check(input) {
			matched = append(matched, r)
		}
	}

	ruleNames := make([]string, len(matched))
	for i, r := range matched {
		ruleNames[i] = fmt.Sprintf("%s: %s", r.ID, r.Name)
	}

	if len(matched) == 0 {
		return PolicyResult{
			Decision:     "PASS",
			MatchedRules: []string{},
			Explanation:  "Auto-approval mode enabled. Audit passed by default.",
		}
	}

	sort.Slice(matched, func(i, j int) bool {
		return matched[i].Priority < matched[j].Priority
	})

	topRule := matched[0]

	return PolicyResult{
		Decision:     "PASS",
		MatchedRules: ruleNames,
		Explanation:  fmt.Sprintf("Auto-approval mode enabled. Audit passed by default. Matched %d informational rule(s). Top rule: %s - %s.", len(matched), topRule.ID, topRule.Name),
	}
}

// containsSignal checks whether a slice of signals contains a specific signal.
func containsSignal(signals []string, target string) bool {
	for _, s := range signals {
		if s == target {
			return true
		}
	}
	return false
}

File: audit/src/lib/audit/policy-engine.ts
import type { AuditDecision } from "@prisma/client";

export interface PolicyInput {
  riskScore: number;
  riskSignals: string[];
  entities: {
    urls: string[];
    qrPayloads: string[];
    riskTerms: string[];
  };
  declaredLandingUrl: string;
  qrUrls: string[];
}

export interface PolicyResult {
  decision: AuditDecision;
  matchedRules: string[];
  explanation: string;
}

interface Rule {
  id: string;
  name: string;
  check: (input: PolicyInput) => boolean;
  decision: AuditDecision;
  priority: number; // lower = higher priority
}

const RULES: Rule[] = [
  {
    id: "R001",
    name: "QR URL mismatch with declared landing page",
    check: (input) => {
      if (input.qrUrls.length === 0) return false;
      try {
        const declaredDomain = new URL(input.declaredLandingUrl).hostname;
        return input.qrUrls.some((url) => {
          try {
            return new URL(url).hostname !== declaredDomain;
          } catch {
            return true;
          }
        });
      } catch {
        return false;
      }
    },
    decision: "REJECT",
    priority: 1,
  },
  {
    id: "R002",
    name: "Contains false profit guarantees",
    check: (input) =>
      input.riskSignals.includes("risk_terms_detected") &&
      input.entities.riskTerms.some((t) =>
        ["guaranteed returns", "100% profit", "risk-free"].includes(t)
      ),
    decision: "REJECT",
    priority: 1,
  },
  {
    id: "R003",
    name: "Suspicious redirect chain detected",
    check: (input) => input.riskSignals.includes("suspicious_redirect"),
    decision: "REJECT",
    priority: 2,
  },
  {
    id: "R004",
    name: "High risk domain detected",
    check: (input) => input.riskSignals.includes("high_risk_domain"),
    decision: "REJECT",
    priority: 2,
  },
  {
    id: "R005",
    name: "QR code present - needs review",
    check: (input) => input.riskSignals.includes("qr_code_found"),
    decision: "MANUAL_REVIEW",
    priority: 5,
  },
  {
    id: "R006",
    name: "Short link usage detected",
    check: (input) => input.riskSignals.includes("short_link_detected"),
    decision: "MANUAL_REVIEW",
    priority: 5,
  },
  {
    id: "R007",
    name: "Telegram handle does not match project",
    check: (input) => input.riskSignals.includes("telegram_mismatch"),
    decision: "MANUAL_REVIEW",
    priority: 6,
  },
  {
    id: "R008",
    name: "Contains wallet connect / claim language",
    check: (input) =>
      input.entities.riskTerms.some((t) =>
        ["connect wallet", "wallet connect", "claim", "claim now"].includes(t)
      ),
    decision: "MANUAL_REVIEW",
    priority: 7,
  },
];

export function evaluatePolicy(input: PolicyInput): PolicyResult {
  const matchedRules: Rule[] = [];

  for (const rule of RULES) {
    if (rule.check(input)) {
      matchedRules.push(rule);
    }
  }

  if (matchedRules.length === 0) {
    return {
      decision: "PASS",
      matchedRules: [],
      explanation: "Auto-approval mode enabled. Audit passed by default.",
    };
  }

  matchedRules.sort((a, b) => a.priority - b.priority);
  const topRule = matchedRules[0];

  return {
    decision: "PASS" as AuditDecision,
    matchedRules: matchedRules.map((r) => `${r.id}: ${r.name}`),
    explanation: `Auto-approval mode enabled. Audit passed by default. Matched ${matchedRules.length} informational rule(s). Top rule: ${topRule.id} - ${topRule.name}.`,
  };
}
